lib/kubernetes/client.rb

```diff
@@ -0,0 +1,136 @@
+  # [[src_port_0, dst_port_0], [src_port_1, dst_port_1],...]
+  def lb_desired_ports(svc)
+    svc["spec"]["ports"].map { |port| [port["port"], port["nodePort"]] }
+  end
+
+  def load_balancer_hostname_missing?(svc)
+    svc.dig("status", "loadBalancer", "ingress", "hostname").to_s.empty?
+  end
+
+  def kubectl(cmd)
+    @session.exec!("sudo kubectl --kubeconfig=/etc/kubernetes/admin.conf #{cmd}")
+  end
+
+  def set_load_balancer_host_name(svc, hostname)
+    patch_data = JSON.generate({
+      "status" => {
+        "loadBalancer" => {
+          "ingress" => [{"hostname" => hostname}]
+        }
+      }
+    })
+    kubectl("-n #{svc.dig("metadata", "namespace")} patch service #{svc.dig("metadata", "name")} --type=merge -p '#{patch_data}' --subresource=status")
+  end
+
+  def add_svc_finalizer(svc)
+    patch_data = JSON.generate({
+      "metadata" => {"finalizers" => [SERVICE_FINALIZER]}
+    })
+    kubectl("-n #{svc.dig("metadata", "namespace")} patch service #{svc.dig("metadata", "name")} --type=merge -p '#{patch_data}'")
+  end
+
+  def remove_svc_finalizer(svc)
+    patch_data = JSON.generate([
+      {"op" => "remove", "path" => "/metadata/finalizers", "value" => [SERVICE_FINALIZER]}
+    ])
+    kubectl("-n #{svc.dig("metadata", "namespace")} patch service #{svc.dig("metadata", "name")} --type=json -p='#{patch_data}'")
+  end
+
+  def vm_diff_for_lb(lb)
+    worker_vms = @kubernetes_cluster.nodepools.flat_map(&:vms)
+    worker_vm_ids = worker_vms.map(&:id).to_set
+    lb_vms = lb.load_balancers_vms.map(&:vm)
+    lb_vm_ids = lb_vms.map(&:id).to_set
+
+    extra_vms = lb_vms.reject { |vm| worker_vm_ids.include?(vm.id) }
+    missing_vms = worker_vms.reject { |vm| lb_vm_ids.include?(vm.id) }
+    [extra_vms, missing_vms]
+  end
+
+  def port_diff_for_lb(lb, desired_ports)
+    lb_ports_hash = lb.ports.to_h { |p| [[p.src_port, p.dst_port], p.id] }
+    missing_ports = desired_ports - lb_ports_hash.keys
+    extra_ports = (lb_ports_hash.keys - desired_ports).map { |p| LoadBalancerPort[id: lb_ports_hash[p]] }
+
+    [extra_ports, missing_ports]
+  end
+
+  def lb_service_modified?(svc)
+    return true if is_service_deleted(svc) && is_service_finalized(svc)
+
+    lb = LoadBalancer.where(name: lb_name_for_svc).first
+    return true unless lb
+
+    extra_vms, missing_vms = vm_diff_for_lb(lb)
+    return true unless extra_vms.empty? && missing_vms.empty?
+
+    extra_ports, missing_ports = port_diff_for_lb(lb, lb_desired_ports(svc))
+    return true unless extra_ports.empty? && missing_ports.empty?
+
+    load_balancer_hostname_missing?(svc)
+  end
+
+  def any_lb_services_modified?
+    k8s_svc_raw = kubectl("get service --all-namespaces --field-selector spec.type=LoadBalancer -ojson")
+    found = JSON.parse(k8s_svc_raw)["items"].find { |svc| lb_service_modified?(svc) }
+    !!found
+  end
+
+  def reconcile_kubernetes_service_deletion(svc)
+    LoadBalancer.where(name: lb_name_for_svc).first&.incr_destroy
```

Comment:
I didn't get this part, don't we use a single Lb for multiple services? If so, if the service is dropped, why do we drop the LB?